File: module-1/simple_blockchain.rb
require 'digest'  # Import the Digest library for SHA-256 hashing
require 'json'    # Import the JSON library for pretty-printing the blockchain

class Block
  attr_reader :index, :timestamp, :data, :previous_hash, :hash

  # Constructor for creating a new block
  def initialize(index, timestamp, data, previous_hash = '')
    @index = index                 # Block's index in the blockchain
    @timestamp = timestamp         # Timestamp of when the block is created
    @data = data                   # Data stored in the block (e.g., transactions)
    @previous_hash = previous_hash # Hash of the previous block in the chain
    @hash = calculate_hash         # Hash of the current block
  end

  # Calculate the hash of the block
  def calculate_hash
    sha = Digest::SHA256.new    # Create a new SHA-256 hash object
    sha.update("#{index}#{timestamp}#{data}#{previous_hash}")  # Update the hash with block data
    sha.hexdigest              # Get the hexadecimal representation of the hash
  end

  # Setter method for updating the previous_hash attribute
  def previous_hash=(value)
    @previous_hash = value
  end
end

class Blockchain
  attr_reader :chain

  # Initialize the blockchain with a genesis block
  def initialize
    @chain = [create_genesis_block]
  end

  # Create the genesis block, the first block in the blockchain
  def create_genesis_block
    Block.new(0, '01/01/2022', 'Genesis Block', '0')  # Genesis block has no previous hash
  end

  # Get the latest block in the blockchain
  def get_latest_block
    @chain.last
  end

  # Add a new block to the blockchain with provided data
  def add_block(data)
    # Create a new block with an incremented index, current timestamp, provided data,
    # and the hash of the latest block as the previous hash
    new_block = Block.new(@chain.length, Time.now.to_s, data, get_latest_block.hash)
    @chain << new_block  # Append the new block to the blockchain
  end

  # Check if the blockchain is valid by verifying hashes and previous hash links
  def is_chain_valid
    (1...@chain.length).each do |i|
      current_block = @chain[i]
      previous_block = @chain[i - 1]

      # Verify that the current block's hash matches the calculated hash
      return false unless current_block.hash == current_block.calculate_hash

      # Verify that the previous_hash in the current block matches the hash of the previous block
      return false unless current_block.previous_hash == previous_block.hash
    end

    true  # If all checks pass, the blockchain is considered valid
  end
end

# Create a new blockchain
my_blockchain = Blockchain.new

# Add some blocks to the blockchain
my_blockchain.add_block({ amount: 4 })  # Adding a block with data
my_blockchain.add_block({ amount: 8 })  # Adding another block with data

# Check if the blockchain is valid and print the result
puts 'Blockchain is valid: ' + my_blockchain.is_chain_valid.to_s

# Uncomment the lines below to attempt tampering with data (preventing this by making the block immutable)
# my_blockchain.chain[1].data = { amount: 1000 }
# my_blockchain.chain[1].hash = my_blockchain.chain[1].calculate_hash

# Print the blockchain in a JSON format
puts JSON.pretty_generate(my_blockchain.chain)

File: module-1/SimpleBlockchain.java
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Date;

class Block {
    private int index;
    private long timestamp;
    private String data;
    private String previousHash;
    private String hash;

    // Constructor to create a new block
    public Block(int index, String data, String previousHash) {
        this.index = index;
        this.timestamp = new Date().getTime();
        this.data = data;
        this.previousHash = previousHash;
        this.hash = calculateHash();
    }

    // Calculate the hash of the block using SHA-256
    public String calculateHash() {
        String dataToHash = index + timestamp + data + previousHash;
        StringBuilder hash = new StringBuilder();

        try {
            MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
            byte[] hashedBytes = messageDigest.digest(dataToHash.getBytes());

            // Convert bytes to hexadecimal representation
            for (byte hashedByte : hashedBytes) {
                String hex = Integer.toHexString(0xff & hashedByte);

                if (hex.length() == 1) {
                    hash.append('0'); // Ensure leading zeros
                }

                hash.append(hex);
            }
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }

        return hash.toString();
    }

    // Getter methods for block attributes
    public int getIndex() {
        return index;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public String getData() {
        return data;
    }

    public String getPreviousHash() {
        return previousHash;
    }

    public String getHash() {
        return hash;
    }

    // Setter method for updating the hash attribute (used for tampering prevention)
    public void setHash(String hash) {
        this.hash = hash;
    }

    // Setter method for updating the data attribute (used for tampering prevention)
    public void setData(String data) {
        this.data = data;
    }
}

class Blockchain {
    private ArrayList<Block> chain;

    // Constructor to initialize the blockchain with a genesis block
    public Blockchain() {
        chain = new ArrayList<>();
        createGenesisBlock();
    }

    // Create the genesis block and add it to the blockchain
    private void createGenesisBlock() {
        chain.add(new Block(0, "Genesis Block", "0"));
    }

    // Get the latest block in the blockchain
    public Block getLatestBlock() {
        return chain.get(chain.size() - 1);
    }

    // Add a new block to the blockchain with provided data
    public void addBlock(String data) {
        Block previousBlock = getLatestBlock();
        int newIndex = previousBlock.getIndex() + 1;
        String newHash = calculateHash(newIndex, previousBlock.getHash(), data);
        chain.add(new Block(newIndex, data, newHash));
    }

    // Calculate the hash for a new block based on its index, previous hash, and data
    private String calculateHash(int index, String previousHash, String data) {
        String dataToHash = index + previousHash + data;
        StringBuilder hash = new StringBuilder();

        try {
            MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
            byte[] hashedBytes = messageDigest.digest(dataToHash.getBytes());

            // Convert bytes to hexadecimal representation
            for (byte hashedByte : hashedBytes) {
                String hex = Integer.toHexString(0xff & hashedByte);

                if (hex.length() == 1) {
                    hash.append('0'); // Ensure leading zeros
                }

                hash.append(hex);
            }
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }

        return hash.toString();
    }

    // Check if the blockchain is valid by verifying hashes and previous hash links
    public boolean isChainValid() {
        for (int i = 1; i < chain.size(); i++) {
            Block currentBlock = chain.get(i);
            Block previousBlock = chain.get(i - 1);

            // Verify that the current block's hash matches the calculated hash
            if (!currentBlock.getHash().equals(currentBlock.calculateHash())) {
                return false;
            }

            // Verify that the previous hash in the current block matches the hash of the previous block
            if (!currentBlock.getPreviousHash().equals(previousBlock.getHash())) {
                return false;
            }
        }

        return true;
    }
}

// Main class to run the blockchain
public class SimpleBlockchain {
    public static void main(String[] args) {
        Blockchain blockchain = new Blockchain();

        // Add blocks to the blockchain
        blockchain.addBlock("Transaction 1");
        blockchain.addBlock("Transaction 2");

        // Check if the blockchain is valid and print the result
        System.out.println("Blockchain is valid: " + blockchain.isChainValid());

        // Uncomment the lines below to attempt tampering with data (preventing this by making the block immutable)
        // Block blockToTamper = blockchain.getLatestBlock();
        // blockToTamper.setData("Tampered Data");
        // blockToTamper.setHash(blockToTamper.calculateHash());

        // Check the validity again after tampering
        // System.out.println("Blockchain is valid: " + blockchain.isChainValid());
    }
}
